app/models/manageiq/providers/redhat/inventory/persister/target_collection.rb

```diff
@@ -117,6 +89,31 @@ def add_vms_inventory_collections(manager_refs)
         :strategy => :local_db_find_missing_references
       )
     )
+    add_inventory_collection(
+      infra.ems_folder_children(
+        :dependency_attributes => {
+          :clusters    => [collections[:ems_clusters]],
+          :datacenters => [collections[:datacenters]],
```

Comment:
placed here, this will result to nil and to the error 'no method .saved? for nil Class' we saw

Reply:
The issue is different. The code that I moved is related to vm only and it doesn't make sense when we target a host. I think that condition is not enough.

Reply:
right, what I mean is that collections[:clusters] is being add in add_clusters_inventory_collections, so it will be nil here. And the nil will result in error.

So the manual dependencies must be set when all of the collections have been initialized. You should be able to replace ```collections[:ems_clusters]``` with ```self.ems_clusters``` to get early exception

Reply:
In our model it is not possible to have a cluster or datacenter to be `nil`. We will always have at least one.

Reply:
right, but this is about

````
add_inventory_collection(infra.ems_clusters( ...
````

is creating the ```collections[:ems_clusters]``` so trying to access it before will result in nil.

That is why I propose using the ```self.ems_clusters```, you will get no method defined exception, until the ems_clusters InventoryCollection is initialized

Reply:
btw. that is the reason why the moved code was placed after all InventoryCollections were initialized (with the bug that add_remaining_inventory_collections is not initializing all the remaining, so we still get nil)

Reply:
Right, please see the new version. It should be ok now.